add get_name patch/spool method

d-chambers opened the pull request:
## Description

This PR implements a `get_name` method for the patch and the spool to return a series of names. The names are either generated by columns provided to the function (which has sensible defaults to maintain the old behavior) or by existing columns such as `name` or `path`. 

This PR supersede  #461.

## Checklist

I have (if applicable):

- [ ] referenced the GitHub issue this PR closes.
- [ ] documented the new feature with docstrings or appropriate doc page.
- [ ] included a test. See [testing guidelines](https://dascore.org/contributing/testing.html).
- [ ] your name has been added to the contributors page (docs/contributors.md).
- [ ] added the "ready_for_review" tag once the PR is ready to be reviewed.

d-chambers commented:
@ahmadtourei, mind taking this for a spin to see if it behaves how you wanted ?

ahmadtourei commented on dascore/utils/patch.py:
```diff
@@ -446,21 +431,90 @@ def get_start_stop_step(patch: PatchType, dim):
+    use / as a delinater.
 
-    def _format_datetime64(dt):
+    Examples
+    --------
+    >>> import dascore as dc
+    >>> from dascore.utils.patch import get_patch_names
+    >>> patch = dc.get_example_patch()
+    >>> name = get_patch_names(patch)
```
It would be helpful to add a few examples of using this function for directory and memory spools. Based on the current function's name, user may assume it can be only applied on patch.

ahmadtourei commented on dascore/utils/patch.py:
```diff
@@ -446,21 +431,90 @@ def get_start_stop_step(patch: PatchType, dim):
+    use / as a delinater.
 
-    def _format_datetime64(dt):
+    Examples
+    --------
+    >>> import dascore as dc
+    >>> from dascore.utils.patch import get_patch_names
+    >>> patch = dc.get_example_patch()
+    >>> name = get_patch_names(patch)
```
`name` is:

```python
0    DAS_______random__2017_09_18__2017_09_18T00_00_07
Name: network, dtype: object
```

We get `DAS_______random__2017_09_18__2017_09_18T00_00_07` using `name[0]` because the function returns a `<class 'pandas.core.series.Series'>`. Is there a specific reason that it returns a pandas series instead of the name in a string?

d-chambers replied:
Yes, for the function to work with a spool it needs to return multiple names. I avoided returning multiple types (based on if the input is a Patch or not) from `get_patch_names` as that is a bit of an anti-pattern. 

I understand the confusion though. I will add `get_patch_name` as a method of `Patch` that will do this. I think with the different name (`get_patch_name` vs `get_patch_names`) it will be clear.

ahmadtourei commented on dascore/utils/patch.py:
```diff
@@ -446,21 +431,90 @@ def get_start_stop_step(patch: PatchType, dim):
-def get_default_patch_name(patch):
-    """Generates the name of the node."""
+def get_patch_names(
+    patch_data: pd.DataFrame | dc.Patch | dc.BaseSpool,
+    prefix="DAS",
+    attrs=("network", "station", "tag"),
+    coords=("time",),
+    sep="__",
```
need test(s) for optional inputs . For example, get_patch_names(patch, patch) does not raise an error.

ahmadtourei commented on tests/conftest.py:
```diff
@@ -385,6 +385,15 @@ def one_file_dir(tmp_path_factory, random_patch):
     return ex.spool_to_directory(spool, path=out)
 
 
+@pytest.fixture(scope="session")
+def random_spool_directory(tmp_path_factory):
+    """A directory with a few patch files."""
+    out = Path(tmp_path_factory.mktemp("one_file_file_spool"))
+    spool = ex.get_example_spool("random_das")
+    out_path = ex.spool_to_directory(spool, path=out)
+    return dc.spool(out_path).update()
```
It would be nice to have this as an example in dascore.examples as all of the example spools are memory spools.

ahmadtourei reviewed (changes requested):
Looks good! I have posted my comments below. 

One general comment that I have is because this can be a very useful function, it would be great to create a shortcut for it (something like `patch.name` and `spool.name` so the user can quickly get name(s) in string format. Then we can include it in the Shortcuts section of `dascore/docs/tutorial/patch.qmd`. I assume users with limited DASCore experience cannot quickly locate it in `dascore.utils.patch`.

d-chambers commented:
>  I assume users with limited DASCore experience cannot quickly locate it in dascore.utils.patch.

I agree. I think it should be a patch/spool method. Because there are optional parameters though, probably best not to make it an attribute (because then the users are always stuck with the defaults).

d-chambers commented:
Thanks @ahmadtourei for the great review. I have addressed all of your suggestions/concerns. Anything else you want to see in here? I also included a fix for the doc build so I am keen to merge this soon to see if the fix works.

ahmadtourei commented on dascore/examples.py:
```diff
@@ -451,6 +451,24 @@ def random_spool(time_gap=0, length=3, time_min=np.datetime64("2020-01-03"), **k
     return dc.spool(out)
 
 
+@register_func(EXAMPLE_SPOOLS, key="radom_directory_das")
```
*random

ahmadtourei replied:
Also, currently getting the following error trying to getting the spool:
```python
import dascore as dc
sp  = dc.get_example_spool("radom_directory_das")

AttributeError: module 'dascore.io.segy.core' has no attribute 'SegyV2'. Did you mean: 'SegyV2_0'?
```
I see you've already merged master so I'm unsure what the problem is.

ahmadtourei commented on docs/tutorial/patch.qmd:
````diff
@@ -227,6 +227,18 @@ print(patch.channel_count) # to get the number of channels in the patch.
 
 These only work for patches with dimensions "time" and "distance" but can help new users who may be unfamiliar datetimes and coordinates. 
 
+The "name" of the patch, which is the filename that would be used if the patch were saved to disk by DASCore, can be generated via the [`Spool.get_patch_names`](`dascore.BaseSpool.get_patch_names`) to get a series of the names of the managed patches, or [`Patch.get_patch_name`](`dascore.Patch.get_patch_name`) to get a string of the patch name.
+
+```python 
+import dascore as dc
+
+patch = dc.get_example_patch()
+spool = dc.get_example_spool()
+
+patch_name = patch.get_patch_name()
+patch_names = spool.get_patch_names()
+```
````
great!

ahmadtourei commented:
> Thanks @ahmadtourei for the great review. I have addressed all of your suggestions/concerns. Anything else you want to see in here? I also included a fix for the doc build so I am keen to merge this soon to see if the fix works.

Thanks, Derrick, for your work on this PR! It looks great. The only new comment I have is a typo I listed below.
